waveform_editor/derived_waveform.py

```diff
@@ -0,0 +1,170 @@
+        except Exception as e:
+            self.annotations.add(0, f"Could not parse or evaluate the waveform: {e}")
+            self.expression = None
+            return
+
+        extractor = ExpressionExtractor()
+        modified_tree = extractor.visit(ast.fix_missing_locations(tree))
+        self.is_constant = extractor.is_constant
+        self.expression = ast.unparse(modified_tree)
+        self.dependent_waveforms = set(extractor.string_nodes)
+
+    def rename_dependency(self, old_name, new_name):
+        """Rename a dependency waveform in the expression.
+
+        Args:
+            old_name: Original dependency name.
+            new_name: New dependency name.
+        """
+        if old_name not in self.dependent_waveforms:
+            return
+
+        tree = ast.parse(self.yaml, mode="eval")
+        renamer = DependencyRenamer(old_name, new_name, self.yaml)
+        ast.fix_missing_locations(renamer.visit(tree))
+        self.yaml = renamer.yaml
+        self.prepare_expression()
+
+    def _build_eval_context(self, time: np.ndarray) -> dict:
+        """Build the evaluation context dictionary with dependencies resolved.
+
+        Args:
+            time: The time array on which to generate points.
+
+        Returns:
+            dict: Mapping dependency names to waveform values at given times.
+        """
+        eval_context = {}
+
+        for name in self.dependent_waveforms:
+            eval_context[name] = self.config[name].get_value(time)[1]
+        return eval_context
+
+    def get_value(
+        self, time: Optional[np.ndarray] = None
+    ) -> tuple[np.ndarray, np.ndarray]:
+        """Evaluate the derived waveform expression at specified times.
+
+        Args:
+            time: Array of time points. Defaults to 1000 points between config.start
+                and config.end.
+
+        Returns:
+            Tuple containing the time and the derived waveform values.
+        """
+        if time is None:
+            # TODO: properly handle time for plotting
+            time = np.linspace(self.config.start, self.config.end, 1000)
+        if self.expression is None:
+            return time, np.zeros_like(time)
+
+        eval_context = self._build_eval_context(time)
+        aeval = Interpreter(user_symbols={"__w": eval_context}, minimal=True)
+        with np.printoptions(threshold=10):
+            result = aeval.eval(self.expression, raise_errors=True)
+
+        # If derived waveform is a constant, ensure an array is returned
+        if self.is_constant:
+            return time, np.full_like(time, result)
+
+        # Ensure the result is a 1D array
+        if not isinstance(result, (list, np.ndarray)):
+            raise ValueError("Evaluation result is not a list or 1D array")
+        result = np.asarray(result)
+        if result.ndim != 1:
+            raise ValueError("Evaluation result is not 1-dimensional")
+        if result.shape[0] != len(time):
```

Comment:
I'd be happy to check `if result.shape != time.shape` with a single error message (that includes both shapes)